modules/custom/social_magic_login/src/Service/MagicUrlCreate.php

```diff
@@ -0,0 +1,76 @@
+<?php
+
+namespace Drupal\social_magic_login\Service;
+
+use Drupal\Core\Path\PathValidatorInterface;
+use Drupal\Core\Url;
+use Drupal\user\UserInterface;
+
+/**
+ * Class MagicLoginCreate.
+ */
+class MagicUrlCreate {
+
+  /**
+   * The path validator service.
+   *
+   * @var \Drupal\Core\Path\PathValidatorInterface
+   */
+  protected $pathValidator;
+
+  /**
+   * MagicUrlCreate constructor.
+   *
+   * @param \Drupal\Core\Path\PathValidatorInterface $pathValidator
+   *   The path validator.
+   */
+  public function __construct(PathValidatorInterface $pathValidator) {
+    $this->pathValidator = $pathValidator;
+  }
+
+  /**
+   * Create a magic login link.
+   *
+   * @param \Drupal\user\UserInterface $account
+   *   An object containing the user account.
+   * @param string $destination
+   *   The uri of the final destination.
+   * @param array $options
+   *   (optional) A keyed array of settings. Supported options are:
+   *   - langcode: A language code to be used when generating locale-sensitive
+   *    URLs. If langcode is NULL the users preferred language is used.
+   *
+   * @return \Drupal\Core\Url
+   *   An url based on the magic login route.
+   */
+  public function create(UserInterface $account, $destination, array $options) {
```

Comment:
Since this is a new service we can add parameter and return type hints here.

```suggestion
  public function create(UserInterface $account, string $destination, array $options) : ?Url {
```